# Árbol de decisión

Según el siguiente árbol de decisión, analice y responda:

![Obama-Clinton decision tree](https://github.com/justmarkham/DAT8/raw/226791169b1cc6df8e8845c12e34e748d5ffaa85/notebooks/images/obama_clinton_tree.jpg)

## 1.1 ¿Qué son las observaciones? ¿Cuántas observaciones hay?


La división geográfica, administrativa y política de Estados Unidos principalmente se define por 50 estados, los cuales están subdivididos en condados. <BR> Ahora bien, en Estados Unidos, el presidente es elegido mediante un colegio electoral y cada estado tiene una cantidad de votos electorales equivalente al número de senadores y representantes del congreso. La votación de cada miembro del colegio electoral será definida por cómo votaron los ciudadanos. <BR> Por lo tanto, las observaciones del árbol corresponden a condados y se tiene un total de 2241.

## 1.2 ¿Cuál es la variable de respuesta?


La variable respuesta es la posibilidad que tiene un candidado de ganar las elecciones teniendo en cuenta diferentes factores. Esta posibilidad está representada como el chance que tiene cierto candidato de ganar con respecto a la posibilidad que tendría en contra. <BR> En otras palabras, dice cuántos votos favorables tendría contra los votos que no recibiría en ciertos condados.

## 1.3 ¿Cuáles son las características?¿Cuál es la característica más predictiva?


Las características utilizadas para discriminar en el árbol de decisión son:
* Los condados que tengan un porcentaje de población afroamericana mayor al 20%, que se definirían como condados con una población afroamericana grande.
* Los condados que tengan una tasa de graduados de la secundaria mayor al 78%,que se definirían como condados con un mayor número de votantes educados.
* Los condados que tengan una tasa de graduados de la secundaria mayor al 87% que se definirían como condados con un alto nivel educativo, en contra de condados en donde del 78-87% de la población tienen un diploma.
* La ubicación geográfica del condado dependiendo si está ubicado al oeste o al medio oeste; o por otro lado, está ubicado al noreste o al sur de Estados Unidos.
* La cantidad de hogares pobres para el año 2000 que se define mediante el ingreso. Entonces, esta característica va a dictar que habían muchos hogares pobres si al menos el 47% de los hogares tenían ingresos menores a 30.000 dólares anuales, en cambio, si más del 53% de los hogares tenían un ingreso anual superior a 30.000 dólares no eran considerados hogares pobres.
* La densidad de la población, en donde la característica ubicaba por un lado a los condados que tenían más de 61.5 personas por milla cuadrada a un lado y a los que tenían menos de eso, los clasificaba como condados rurales.
* La última característica tenía en cuenta el resultado electoral para el año 2004, en donde si Bush vencía a Kerry en más de 16.5 puntos porcentuales se podía considerar que esos condados eran muy republicanos. <BR> 
Para finalizar, se dice que la característica más predictiva es el porcentaje de afroamericanos que hay en el condado, puesto que es la característica principal que más ayuda a diferenciar.


## 1.4 ¿Cuál es la predicción de clase para el siguiente condado con los siguientes datos: 15% de afroamericanos, 90% de tasa de graduación en la escuela secundaria, ubicado en el Sur, alta pobreza, alta densidad de población?


Según las características dadas, se predice que el condado va a elegir a Obama como el presidente de Estados Unidos para el periodo 2008-2012.

## 1.5 ¿Cuál es el chance predicha para ese mismo condado anteriormente descrito?

Para el condado anteriormente mencionado, se dice que hay una posibilidad de 185 contra 36 de que gane Obama. En otras palabras, Obama tiene un chance de 185 votos de ganar contra 36 votos que no lo eligirían.

# SVM


Para esta tarea consideramos un conjunto de observaciones sobre una serie de variedades de vino tinto y blanco que implican sus propiedades químicas y su clasificación por parte de los catadores. La industria del vino ha experimentado un gran crecimiento en los últimos tiempos debido al aumento del consumo social. El precio del vino depende de un concepto bastante abstracto de apreciación del vino por parte de los catadores, cuya opinión puede tener un alto grado de variabilidad. El precio del vino depende en cierta medida de este factor tan volátil. Otro factor clave en la certificación y evaluación de la calidad del vino son las pruebas fisicoquímicas, que se realizan en laboratorio y tienen en cuenta factores como la acidez, el nivel de pH, la presencia de azúcar y otras propiedades químicas. Para el mercado del vino, sería interesante que la calidad humana de la cata pudiera relacionarse con las propiedades químicas del vino para que el proceso de certificación y evaluación de la calidad estuviera más controlado.

Se dispone de dos conjuntos de datos, uno de los cuales se refiere al vino tinto y cuenta con 1.599 variedades diferentes, y el otro, al vino blanco, con 4.898 variedades. Todos los vinos se producen en una zona concreta de Portugal. Se recogen datos sobre 12 propiedades diferentes de los vinos, una de las cuales es la calidad, basada en datos sensoriales, y el resto son propiedades químicas de los vinos, como la densidad, la acidez, el contenido de alcohol, etc. Todas las propiedades químicas de los vinos son variables continuas. La calidad es una variable ordinal con una clasificación posible de 1 (peor) a 10 (mejor). Cada variedad de vino es catada por tres catadores independientes y la clasificación final asignada es la mediana de la clasificación dada por los catadores.

Se espera que un modelo predictivo desarrollado a partir de estos datos sirva de orientación a los viñedos en cuanto a la calidad y el precio que se espera obtener de sus productos sin depender en gran medida de la volatilidad de los catadores.


In [15]:
import pandas as pd
import numpy as np

In [25]:
data_r = pd.read_csv('https://raw.githubusercontent.com/prof-tatiana-gutierrez/DatosMD/master/Wine_data_red.csv')
data_w = pd.read_csv('https://raw.githubusercontent.com/prof-tatiana-gutierrez/DatosMD/master/Wine_data_white.csv')

In [35]:
data = data_w.assign(type = 'white')

data = data.append(data_r.assign(type = 'red'), ignore_index=True)
data.sample(5)

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
279             7.0              0.30         0.49             4.7      0.036   
6077            8.2              0.35         0.33             2.4      0.076   
956             7.5              0.32         0.26             1.8      0.042   
1264            7.2              0.14         0.35             1.2      0.036   
1093            8.4              0.17         0.31             6.7      0.038   

      free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
279                  17.0                 105.0  0.99160  3.26       0.68   
6077                 11.0                  47.0  0.99599  3.27       0.81   
956                  13.0                 133.0  0.99380  3.07       0.38   
1264                 15.0                  73.0  0.99380  3.46       0.39   
1093                 29.0                 132.0  0.99450  3.10       0.32   

      alcohol  quality   type  
279      12.4        7  white  
6077     11.0        6    red  
956       9.5        5  white  
1264      9.9        5  white  
1093     10.6        7  white

In [27]:
data.quality.value_counts()

6    2836
5    2138
7    1079
4     216
8     193
3      30
9       5
Name: quality, dtype: int64

In [28]:
data.type.value_counts()

white    4898
red      1599
Name: type, dtype: int64

## Ejercicio 2.1

Mostrar la tabla de frecuencias de la calidad por tipo de vino.

In [29]:
tab=pd.crosstab(data["type"],data["quality"])
tab

quality   3    4     5     6    7    8  9
type                                     
red      10   53   681   638  199   18  0
white    20  163  1457  2198  880  175  5

De acuerdo a la anterior tabla de contingencia se puede decir que:
* Para el vino tinto, 10 catadores dieron una calificación de 3 para la calidad del vino que tomaron, 53 una calificación de 4, 681 una calificación de 5, 638 una calificación de 6, 199 una calificación de 7 y 18 una calificación de 8 a la calidad de la variedad de vino tinto que probaron; para un total de 1599.
* Para el vino blanco, 20 catadores dieron una calificación de 3 para la calidad del vino que tomaron, 163 una calificación de 4, 1457 una calificación de 5, 2198 una calificación de 6, 880 una calificación de 7, 175 una calificación de 8 y 5 una calificación de 9 a la calidad de la variedad de vino blanco que probaron; para un total de 4898.

## Ejercicio 2.2

Teniendo en cuenta todas las variables:

* Crear un objetivo binario para cada tipo de vino, haciendo la transformación necesaria para ingresar la información al modelo.
* Crear una SVM que prediga el tipo de vino
* Evalúe con F1, AUC-ROC y log-loss


### Creación de varible binaria para el tipo de vino:
Se le va a asignar 1 al vino tinto y 0 al vino blanco

In [36]:
data['B_type'] =data['type']
data.loc[data.type == 'red','B_type'] = '1'
data.loc[data.type == 'white','B_type'] = '0'
data['B_type'] = data['B_type'].astype(int)

In [31]:
data.B_type.value_counts()

0    4898
1    1599
Name: B_type, dtype: int64

### Ajuste del modelo:

In [37]:
import numpy as np
import matplotlib.pyplot as plt
from scipy import stats
import seaborn as sns
from sklearn.svm import SVC
from sklearn.datasets import make_blobs
from sklearn import preprocessing, metrics, model_selection
import pandas as pd
from sklearn.metrics import roc_auc_score, accuracy_score, make_scorer, roc_curve, auc
from sklearn.preprocessing import label_binarize
from sklearn.model_selection import train_test_split
from sklearn.model_selection import GridSearchCV
import warnings
warnings.filterwarnings('ignore')

data.drop(['type'], 1, inplace=True)
X = np.array(data.drop(['B_type'], 1))
y = np.array(data['B_type'])
X_train, X_test, y_train, y_test = model_selection.train_test_split(X, y, test_size=0.3)

In [55]:
from sklearn.svm import SVC
clf1 = SVC(kernel='linear', C=1E6, probability= True)
clf1.fit(X_train, y_train)

SVC(C=1000000.0, kernel='linear', probability=True)

### Evaluación del modelo:

In [56]:
pred1 = clf1.predict(X_test)
pred11 = clf1.predict_proba(X_test)
print('F1 score:', metrics.f1_score(y_test, pred1))
print('Log Loss:', metrics.log_loss(y_test, pred11))
print('ROC AUC score:', metrics.roc_auc_score(y_test, pred1, average=None))

F1 score: 0.9678800856531048
Log Loss: 0.04415006818444153
ROC AUC score: 0.975430609389693


## Ejercicio 2.3

* Probar la SVM utilizando los diferentes kernels ('poly', 'rbf', 'sigmoid')
* Evalúe con F1, AUC-ROC y log-loss


### SVM con kernel = poly

In [57]:
from sklearn import svm

clf2 = svm.SVC(kernel='poly',probability=True)
clf2.fit(X_train, y_train)

pred2 = clf2.predict(X_test)
pred22 = clf2.predict_proba(X_test)

print('F1 score:', metrics.f1_score(y_test, pred2))
print('Log Loss:', metrics.log_loss(y_test, pred22))
print('ROC AUC score:', metrics.roc_auc_score(y_test, pred2, average=None))

F1 score: 0.8356309650053022
Log Loss: 0.234511774498065
ROC AUC score: 0.8913241668769064


### SVM con kernel = rbf

In [58]:
clf3 = svm.SVC(kernel='rbf',probability=True)
clf3.fit(X_train, y_train)

pred3 = clf3.predict(X_test)
pred33 = clf3.predict_proba(X_test)

print('F1 score:', metrics.f1_score(y_test, pred3))
print('Log Loss:', metrics.log_loss(y_test, pred33))
print('ROC AUC score:', metrics.roc_auc_score(y_test, pred3, average=None))

F1 score: 0.8664343786295006
Log Loss: 0.20013513147099257
ROC AUC score: 0.8897143987523223


### SVM con kernel = sigmoid

In [59]:
clf4 = svm.SVC(kernel='sigmoid',probability=True)
clf4.fit(X_train, y_train)

pred4 = clf4.predict(X_test)
pred44 = clf4.predict_proba(X_test)

print('F1 score:', metrics.f1_score(y_test, pred4))
print('Log Loss:', metrics.log_loss(y_test, pred44))
print('ROC AUC score:', metrics.roc_auc_score(y_test, pred4, average=None))

F1 score: 0.49946178686759957
Log Loss: 0.485780357271307
ROC AUC score: 0.6696463383867343


### Resumen de evaluaciones:

In [62]:
result = pd.DataFrame({'Modelo':['Modelo SVC','Modelo kernel poly','Modelo kernel rbf','Modelo kernel sigmoid']})
result['F1']=(metrics.f1_score(y_test, pred1),metrics.f1_score(y_test, pred2),metrics.f1_score(y_test, pred3),metrics.f1_score(y_test, pred4))
result['Log loss']=(metrics.log_loss(y_test, pred11),metrics.log_loss(y_test, pred22),metrics.log_loss(y_test, pred33),metrics.log_loss(y_test, pred44))
result['ROC AUC']=(metrics.roc_auc_score(y_test, pred1, average=None),metrics.roc_auc_score(y_test, pred2, average=None),metrics.roc_auc_score(y_test, pred3, average=None),metrics.roc_auc_score(y_test, pred4, average=None))
result

Modelo        F1  Log loss   ROC AUC
0             Modelo SVC  0.967880  0.044150  0.975431
1     Modelo kernel poly  0.835631  0.234512  0.891324
2      Modelo kernel rbf  0.866434  0.200135  0.889714
3  Modelo kernel sigmoid  0.499462  0.485780  0.669646

## Ejercicio 2.4
* Utilizando el mejor SVM del punto anterior encontrar los parámetros que da el mejor rendimiento teniendo en cuenta los siguientes hiperparámetros
'C': [0.1, 1, 10, 100, 1000], 'gamma': [0.01, 0.001, 0.0001]
* Evalúe cada modelo con F1, AUC-ROC y log-loss

Nota: Para los valores de C y gamma, deben realizar todas las posibles combinaciones entre estas dos variables.

### Modelo SVC:

In [66]:
from sklearn import svm

clf5 = svm.SVC()

parameters = {'gamma': [0.01, 0.001, 0.0001], 'C': [0.1, 1, 10, 100, 1000]}
acc_scorer = make_scorer(roc_auc_score)

grid_obj = GridSearchCV(clf5, parameters, scoring=acc_scorer)
grid_obj = grid_obj.fit(X_train, y_train)

clf5_b = grid_obj.best_estimator_

clf5_b.fit(X_train, y_train)

pred5 = clf5_b.predict(X_test)


In [67]:
print('F1 score:', metrics.f1_score(y_test, pred5))
print('Log Loss:', metrics.log_loss(y_test, pred5))
print('ROC AUC score:', metrics.roc_auc_score(y_test, pred5, average=None))

F1 score: 0.9766454352441613
Log Loss: 0.3896723470063276
ROC AUC score: 0.9839051856608794


### Modelo kernel rbf:

In [49]:
clf6 = svm.SVC(kernel='rbf',probability=True)

parameters = {'gamma': [0.01, 0.001, 0.0001], 'C': [0.1, 1, 10, 100, 1000]}
acc_scorer = make_scorer(roc_auc_score)

grid_obj = GridSearchCV(clf6, parameters, scoring=acc_scorer)
grid_obj = grid_obj.fit(X_train, y_train)

clf6_b = grid_obj.best_estimator_

clf6_b.fit(X_train, y_train)

pred6 = clf6_b.predict(X_test)
pred66 = clf6_b.predict_proba(X_test)

In [50]:
print('F1 score:', metrics.f1_score(y_test, pred6))
print('Log Loss:', metrics.log_loss(y_test, pred66))
print('ROC AUC score:', metrics.roc_auc_score(y_test, pred6, average=None))

F1 score: 0.9766454352441613
Log Loss: 0.3896723470063276
ROC AUC score: 0.9839051856608794


## Ejercicio 2.5

Compare los resultados con todos los anteriores modelos y eliga ¿Cuál algoritmo con cuáles hiperparámetros es el que tiene mejor rendimiento según la puntuación F1, AUC-ROC y log-loss?

In [53]:
result_final = pd.DataFrame({'Modelo':['Modelo SVC','Modelo kernel poly','Modelo kernel rbf','Modelo kernel sigmoid','Modelo SVC grilla','Modelo kernel rbf grilla']})
result_final['F1']=(metrics.f1_score(y_test, pred1),metrics.f1_score(y_test, pred2),metrics.f1_score(y_test, pred3),metrics.f1_score(y_test, pred4),metrics.f1_score(y_test, pred5),metrics.f1_score(y_test, pred6))
result_final['Log loss']=(metrics.log_loss(y_test, pred1),metrics.log_loss(y_test, pred2),metrics.log_loss(y_test, pred3),metrics.log_loss(y_test, pred4),metrics.log_loss(y_test, pred5),metrics.log_loss(y_test, pred5))
result_final['ROC AUC']=(metrics.roc_auc_score(y_test, pred1, average=None),metrics.roc_auc_score(y_test, pred2, average=None),metrics.roc_auc_score(y_test, pred3, average=None),metrics.roc_auc_score(y_test, pred4, average=None),metrics.roc_auc_score(y_test, pred5, average=None),metrics.roc_auc_score(y_test, pred6, average=None))
result_final

Modelo        F1  Log loss   ROC AUC
0                Modelo SVC  0.967880  0.531370  0.975431
1        Modelo kernel poly  0.835631  2.745421  0.891324
2         Modelo kernel rbf  0.866434  2.036909  0.889714
3     Modelo kernel sigmoid  0.499462  8.236262  0.669646
4         Modelo SVC grilla  0.976645  0.389672  0.983905
5  Modelo kernel rbf grilla  0.976645  0.389672  0.983905

In [68]:
clf6_b

SVC(C=1000, gamma=0.001)

Se observa que los dos mejores modelos se obtuvieron mediante el método de grilla. <BR> Ahora bien, teniendo en cuenta los métodos de kernel que se utilizaron, se llega a la conclusión de que el mejor modelo es un kernel rbf con hiperparámetros:
* C=1000
* gamma=0.001
